### PR TITLE
hammer: librbd: request exclusive lock if current owner cannot execute op

### PR DESCRIPTION
**Pull request by trociny**
http://tracker.ceph.com/issues/17068

### CODE REVIEW

**Comment by trociny** [comment]
@dillaman I have temporary added a hack to test this feature (will be removed after the review).

Also, I am not very happy with my complete_request function: accessing AioCompletion members looks too hackish. May be you see a better option?

File: src/librbd/ImageWatcher.cc [hunk 1]
```diff
@@ -896,6 +896,8 @@ bool ImageWatcher::handle_payload(const FlattenPayload &payload,
 
 bool ImageWatcher::handle_payload(const ResizePayload &payload,
 				  C_NotifyAck *ack_ctx) {
+  return ImageWatcher::handle_payload(UnknownPayload(), ack_ctx);
```
Comment (dillaman) [inline src/librbd/ImageWatcher.cc]: I assume this is the hack?

File: src/librbd/internal.cc [hunk 2]
```diff
@@ -113,6 +121,18 @@ int invoke_async_request(ImageCtx *ictx, const std::string& request_type,
         }
 
         r = remote_request();
+        if (r == -EOPNOTSUPP && request_lock) {
+          AioCompletion *comp = aio_create_completion();
+          ictx->image_watcher->request_lock(boost::bind(&complete_request, _1),
```
Comment (dillaman) [inline src/librbd/internal.cc]: Can't this section just be:

```
AioCompletion *comp = aio_create_completion();
comp->add_request();
comp->finish_adding_requests(ictx->cct);
ictx->image_watcher->request_lock(boost::bind(&AioCompletion::complete_request, _1, ictx->cct, 0));
ictx->owner_lock.put_read();
r = comp->wait_for_complete();
...
```

File: src/librbd/ImageWatcher.cc (hunk 1 above)
Comment (trociny) [inline src/librbd/ImageWatcher.cc]: Yes, just for testing: make remote return -EOPNOTSUPP. It is not going to be committed.

File: src/librbd/internal.cc (hunk 2 above)
Comment (trociny) [inline src/librbd/internal.cc]: @dillaman It looks like it works! Thanks.

Comment (trociny) [inline src/librbd/internal.cc]: Except one issue -- noise in the log:

2016-11-16 15:27:20.087375 7f02db7fe700 -1 librbd::AioCompletion: completed invalid aio_type: 4

@dillaman May I update AioCompletion::complete not to log error for AIO_TYPE_NONE?

Comment (dillaman) [inline src/librbd/internal.cc]: ... or just add the new AIO type for request lock.

Comment (trociny) [inline src/librbd/internal.cc]: @dillaman Is it worth adding a new type on hammer branch? Also, additionally I would need to call comp->init_time(...) to set the type. If you think it is ok, I can do this way though.

Comment (dillaman) [inline src/librbd/internal.cc]: OK, in that case, it's not worth it.

**Comment by trociny** [comment]
@dillaman Updated.

**Comment by dillaman** [comment]
lgtm

**Comment by smithfarm** [comment]
@trociny @dillaman This PR was included in this rbd suite: http://tracker.ceph.com/issues/17151#note-12

There is a reproducible qemu-iotests failure in this run: http://tracker.ceph.com/issues/10773 

Can you take a look? If you think the qemu-iotests failure is unrelated, this PR could be merged.

**Comment by dillaman** [comment]
@smithfarm lgtm

**Pull request merged**